core/services/ocr2/plugins/mercury/plugin.go

```
@@ -31,42 +37,93 @@ func NewServices(
-	ds := mercury.NewDataSource(
-		orm,
-		pipelineRunner,
-		jb,
-		*jb.PipelineSpec,
-		lggr,
-		runResults,
-		chEnhancedTelem,
-		chainHeadTracker,
-		ocr2Provider.ContractTransmitter(),
-		pluginConfig.InitialBlockNumber.Ptr(),
-		feedID,
-	)
-	argsNoPlugin.MercuryPluginFactory = relaymercury.NewFactory(
-		ds,
-		lggr,
-		ocr2Provider.OnchainConfigCodec(),
-		ocr2Provider.ReportCodec(),
-	)
+
+	version, err := decoder.DecodeSchemaVersionFromFeedId(feedID)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	switch version {
+	case 1:
+		ds := mercuryv1.NewDataSource(
+			orm,
+			pipelineRunner,
+			jb,
+			*jb.PipelineSpec,
+			lggr,
+			runResults,
+			chEnhancedTelem,
+			chainHeadTracker,
+			ocr2Provider.ContractTransmitter(),
+			pluginConfig.InitialBlockNumber.Ptr(),
+			feedID,
+		)
+		argsNoPlugin.MercuryPluginFactory = relaymercuryv1.NewFactory(
+			ds,
+			lggr,
+			ocr2Provider.OnchainConfigCodec(),
+			ocr2Provider.ReportCodecV1(),
+		)
+	case 2:
+		ds := mercuryv2.NewDataSource(
+			pipelineRunner,
+			jb,
+			*jb.PipelineSpec,
+			feedID,
+			lggr,
+			runResults,
+			chEnhancedTelem,
+			ocr2Provider.ContractTransmitter(),
+			types.FeedID(*pluginConfig.LinkFeedID),
+			types.FeedID(*pluginConfig.NativeFeedID),
+		)
+		argsNoPlugin.MercuryPluginFactory = relaymercuryv2.NewFactory(
+			ds,
+			lggr,
+			ocr2Provider.OnchainConfigCodec(),
+			ocr2Provider.ReportCodecV2(),
+		)
+	case 3:
+		ds := mercuryv3.NewDataSource(
+			pipelineRunner,
+			jb,
+			*jb.PipelineSpec,
+			feedID,
+			lggr,
+			runResults,
+			chEnhancedTelem,
+			ocr2Provider.ContractTransmitter(),
+			types.FeedID(*pluginConfig.LinkFeedID),
+			types.FeedID(*pluginConfig.NativeFeedID),
+		)
+		argsNoPlugin.MercuryPluginFactory = relaymercuryv3.NewFactory(
+			ds,
+			lggr,
+			ocr2Provider.OnchainConfigCodec(),
+			ocr2Provider.ReportCodecV3(),
+		)
+	default:
+		panic("unknown schema version")
```

Why panic instead of returning an error?

---

No particular reason here, just wanted to make sure we don't run with the unsupported version. Changing to returning error instead